### PR TITLE
Integration with Chroma

### PR DESCRIPTION
**Pull request by kuraleta**

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/Metadata.java [hunk 1]
```diff
@@ -44,6 +44,10 @@ public Metadata copy() {
         return new Metadata(new HashMap<>(metadata));
     }
 
+    public Map<String, String> asMap() {
+        return metadata;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/document/Metadata.java]: please return copy of the map:
`new HashMap<>(metadata)`

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import retrofit2.http.POST;
+import retrofit2.http.Path;
+
+import java.util.List;
+
+interface ChromaApi {
+
+    @GET("/api/v1/collections")
+    @Headers({"Content-Type: application/json"})
+    Call<List<CollectionCreationResponse>> getCollections();
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: please rename CollectionCreationResponse into Collection

Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: minor: please avoid using get in method names

Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: please remove all unused methods

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                .readTimeout(timeout)
+                .writeTimeout(timeout)
+                .build();
+
+        Gson gson = new GsonBuilder()
+                .setFieldNamingPolicy(LOWER_CASE_WITH_UNDERSCORES)
+                .create();
+
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(urlBase)
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: nitpick: urlBase -> baseUrl

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+     * @return The converted score.
+     */
+    private static double distanceToScore(double distance) {
+        if (distance < 0 || distance > 2) {
+            throw new IllegalArgumentException("Distance must be in the range [0, 2]");
+        }
+        return 1 - (distance / 2);
+    }
+
+    private static String generateRandomId(Embedding embedding) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: please use existing  Utils.randomUUID() method

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+        return ids;
+    }
+
+    private void addInternal(String id, Embedding embedding, TextSegment textSegment) {
+        addAllInternal(singletonList(id), singletonList(embedding), textSegment == null ? null : singletonList(textSegment));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> textSegments) {
+        EmbeddingsRequest embeddingsRequest = EmbeddingsRequest.builder()
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: please rename to AddEmbeddingsRequest

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    @Headers({"Content-Type: application/json"})
+    Call<CollectionCreationResponse> getCollection(@Path("collection_name") String collectionName);
+
+    @POST("/api/v1/collections")
+    @Headers({"Content-Type: application/json"})
+    Call<CollectionCreationResponse> createCollection(@Body CollectionCreationRequest collectionCreationRequest);
+
+    @POST("/api/v1/collections/{collection_id}/add")
+    @Headers({"Content-Type: application/json"})
+    Call<Boolean> addEmbeddingToCollection(@Path("collection_id") String collectionId, @Body EmbeddingsRequest embedding);
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: addEmbedding**_s_**

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults) {
+        return findRelevant(referenceEmbedding, maxResults, 0);
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        float[] vector = referenceEmbedding.vector();
+        QueryRequest queryRequest = QueryRequest.builder()
+                .queryEmbedding(singletonList(IntStream.range(0, vector.length)
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: referenceEmbedding.vectorAsList()

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                        .mapToObj(i -> vector[i])
+                        .collect(toList())))
+                .nResults(maxResults)
+                .build();
+
+        SuccessfulResponse nearestNeighbors = chromaClient.getNearestNeighbors(collectionId, queryRequest);
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: Please rename SuccessfulResponse into QueryResponse

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package dev.langchain4j.store.embedding.chroma;
+
+public class ResponseValidator {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java]: pls make it package-private

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * Validates the given SuccessfulResponse object.
+     *
+     * @param response The SuccessfulResponse object to validate.
+     * @throws IllegalArgumentException if the validation fails.
+     */
+    public static void validate(SuccessfulResponse response) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java]: why do we need this? don't we trust chroma?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+                    nearestNeighbors.getDocuments().get(0).get(i) == null ? null : TextSegment.from(nearestNeighbors.getDocuments().get(0).get(i),
+                            nearestNeighbors.getMetadatas().get(0).get(i) == null ? null : new Metadata(nearestNeighbors.getMetadatas().get(0).get(i)))
+            );
+            embeddingMatches.add(textSegmentEmbeddingMatch);
+        }
+        return embeddingMatches;
+    }
+
+    /**
+     * By default, cosine distance will be used. For details: <a href="https://github.com/nmslib/hnswlib/tree/master#python-bindings"></a>
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: https://docs.trychroma.com/usage-guide

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+     * Converts a cosine distance in the range [0, 2] to a score in the range [1, 0].
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: [1, 0] -> [0, 1]

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+     *
+     * @param distance The distance value.
+     * @return The converted score.
+     */
+    private static double distanceToScore(double distance) {
+        if (distance < 0 || distance > 2) {
+            throw new IllegalArgumentException("Distance must be in the range [0, 2]");
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: after you hardcoded distance metric during collection creation, does it return something out of range [0,2]?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    private static List<EmbeddingMatch<TextSegment>> toEmbeddingMatches(SuccessfulResponse nearestNeighbors) {
+        List<EmbeddingMatch<TextSegment>> embeddingMatches = new ArrayList<>();
+        for (int i = 0; i < nearestNeighbors.getIds().get(0).size(); i++) {
+            EmbeddingMatch<TextSegment> textSegmentEmbeddingMatch = new EmbeddingMatch<>(
+                    distanceToScore(nearestNeighbors.getDistances().get(0).get(i)),
+                    nearestNeighbors.getIds().get(0).get(i),
+                    Embedding.from(nearestNeighbors.getEmbeddings().get(0).get(i)),
+                    nearestNeighbors.getDocuments().get(0).get(i) == null ? null : TextSegment.from(nearestNeighbors.getDocuments().get(0).get(i),
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: please refactor, it is hard to understand

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+class CollectionCreationRequest {
+
+    private final String name;
+    private final Map<String, String> metadata;
+
+    /**
+     * Currently, cosine distance is always used as the distance method for chroma implementation
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: please move/copy this comment into ChromaEmbeddingStore

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    private CollectionCreationRequest(Builder builder) {
+        this.name = builder.name;
+        //TODO other distances? Currently, only cosine distance is possible
+        HashMap<String, String> metadata = new HashMap<>();
+        metadata.put("hnsw:space", "cosine");
+        this.metadata = metadata;
+    }
+
+    @Override
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: no need in this

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    public String toString() {
+        return "CollectionCreationRequest{" +
+                "name='" + name + '\'' +
+                ", chromaMetadata=" + metadata +
+                '}';
+    }
+
+    public static Builder builder() {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: no need in builder, this is an internal class that is dreated in one place. please remove and use ctor

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationResponse.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    public String id() {
+        return id;
+    }
+
+    public Map<String, String> metadata() {
+        return metadata;
+    }
+
+    @Override
+    public String toString() {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationResponse.java]: no need in this

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
+package dev.langchain4j.store.embedding.chroma;
+
+import java.util.List;
+
+import static java.util.Arrays.asList;
+
+class QueryRequest {
+
+    private final List<List<Float>> query_embeddings;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: please use camelCase

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    private final List<String> include;
+
+    public QueryRequest(Builder builder) {
+        this.query_embeddings = builder.queryEmbedding;
+        this.n_results = builder.nResults;
+        this.include = builder.include;
+    }
+
+    @Override
+    public String toString() {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: no need in this

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private List<List<Float>> queryEmbedding;
+        private int nResults = 10;
+        private List<String> include = asList("metadatas", "documents", "distances", "embeddings");
+
+        public Builder queryEmbedding(List<List<Float>> queryEmbedding) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: please change this method to accept List<Float> instead

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+class SuccessfulResponse {
+
+    private List<List<String>> ids;
+    private List<List<Double>> distances;
+    private List<List<Map<String, String>>> metadatas;
+    private List<List<List<Float>>> embeddings;
+    private List<List<String>> documents;
+
+    public List<List<String>> getIds() {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java]: please remove get

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    public List<List<List<Float>>> getEmbeddings() {
+        return embeddings;
+    }
+
+    public List<List<String>> getDocuments() {
+        return documents;
+    }
+
+    @Override
+    public String toString() {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java]: not needed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+    public AddEmbeddingsRequest(Builder builder) {
+        this.embeddings = builder.embeddings;
+        this.metadatas = builder.chromaMetadata;
+        this.documents = builder.documents;
+        this.ids = builder.ids;
+        this.incrementIndex = builder.incrementIndex;
+    }
+
+    @Override
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java]: not needed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java (hunk 2 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java (hunk 3 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: ok, done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java (hunk 6 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 12 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 8 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 7 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: good point, changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 5 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java (hunk 2 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: ok

Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: done

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/Metadata.java (hunk 1 above)
Comment (kuraleta) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/Metadata.java]: makes sense

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java (hunk 22 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java (hunk 23 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/SuccessfulResponse.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java (hunk 20 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java (hunk 19 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationResponse.java (hunk 18 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationResponse.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java (hunk 17 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java (hunk 16 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 4 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 13 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 11 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java (hunk 21 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java (hunk 24 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java]: deleted

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java (hunk 9 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java]: deleted

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java (hunk 10 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ResponseValidator.java]: I will try to trust Chroma :) deleted

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 14 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: changed, but I'm not sure if it's better now

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java (hunk 15 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: added comment to ChromaEmbeddingStore

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+class AddEmbeddingsRequest {
+
+    private final List<float[]> embeddings;
+    private final List<Map<String, String>> metadatas;
+    private final List<String> documents;
+    private final List<String> ids;
+
+    public AddEmbeddingsRequest(Builder builder) {
+        this.embeddings = builder.embeddings;
+        this.metadatas = builder.chromaMetadata;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java]: minor: why different names?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+
+interface ChromaApi {
+
+    @GET("/api/v1/collections/{collection_name}")
+    @Headers({"Content-Type: application/json"})
+    Call<Collection> collection(@Path("collection_name") String collectionName);
+
+    @POST("/api/v1/collections")
+    @Headers({"Content-Type: application/json"})
+    Call<Collection> createCollection(@Body CollectionCreationRequest collectionCreationRequest);
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: minor: maybe better to rename CollectionCreationRequest into CreateCollectionRequest?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+            } else {
+                // if collection is not present, Chroma returns: Status - 500
+                return null;
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    boolean addEmbedding(String collectionId, AddEmbeddingsRequest addEmbeddingsRequest) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: addEmbeddingS

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+                return Boolean.TRUE.equals(retrofitResponse.body());
+            } else {
+                throw toException(retrofitResponse);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    QueryResponse getNearestNeighbors(String collectionId, QueryRequest queryRequest) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: maybe queryCollection?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 29]
```diff
@@ -0,0 +1,166 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Represents a store for embeddings with Chroma backend.
+ * The cosine distance is always used as the distance method for Chroma implementation
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: The cosine distance is always used as the distance metric.

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 30]
```diff
@@ -0,0 +1,166 @@
...
+    private final String collectionId;
+
+    /**
+     * Initializes a new instance of ChromaEmbeddingStore with the specified parameters.
+     *
+     * @param urlBase        The base URL of the Chroma service.
+     * @param timeout        The timeout duration for the Chroma client.
+     * @param collectionName The name of the collection in the Chroma service.
+     */
+    public ChromaEmbeddingStore(String urlBase, Duration timeout, String collectionName) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: urlBase -> baseUrl

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 31]
```diff
@@ -0,0 +1,166 @@
...
+        this.chromaClient = new ChromaClient(urlBase, timeout);
+
+        Collection response = chromaClient.getCollection(collectionName);
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: minor: response -> collection

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 32]
```diff
@@ -0,0 +1,166 @@
...
+
+        collectionName = collectionName == null ? "default" : collectionName;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: this line probably should be above "chromaClient.getCollection(collectionName);" ?

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java [hunk 33]
```diff
@@ -0,0 +1,166 @@
...
+            Collection collection = chromaClient.createCollection(new CollectionCreationRequest(collectionName));
+            collectionId = collection.id();
+        } else {
+            collectionId = response.id();
+        }
+    }
+
+    @Override
+    public String add(Embedding embedding) {
+        String id = Utils.randomUUID();
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: minor: please use static imports when class name does not clarify or add relevant information

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 30 above)
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: minor: i would put collectionName on the second place, make timeout optional and add a builder

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java [hunk 34]
```diff
@@ -0,0 +1,22 @@
...
+    private final String name;
+    private final Map<String, String> metadata;
+
+    /**
+     * Currently, cosine distance is always used as the distance method for chroma implementation
+     */
+    CollectionCreationRequest(String name) {
+        this.name = name;
+        HashMap<String, String> metadata = new HashMap<>();
+        //Currently, only cosine distance is possible
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: minor: possible -> supported

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryResponse.java [hunk 35]
```diff
@@ -0,0 +1,34 @@
+package dev.langchain4j.store.embedding.chroma;
+
+import java.util.List;
+import java.util.Map;
+
+class QueryResponse {
+
+    private List<List<String>> ids;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryResponse.java:8]: minor: please have a consistent ordering of parameters in all places (e.g. AddEmbeddingsRequest)

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java [hunk 36]
```diff
@@ -0,0 +1,19 @@
...
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+
+class QueryRequest {
+
+    private final List<List<Float>> queryEmbeddings;
+    private final int nResults;
+    private final List<String> include = asList("metadatas", "documents", "distances", "embeddings");
+
+    public QueryRequest(List<Float> queryEmbeddings, int nResults) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: minor: queryEmbeddings -> queryEmbedding

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 32 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: sure

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 33 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 29 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java (hunk 28 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java (hunk 27 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaClient.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java (hunk 26 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaApi.java]: changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java (hunk 25 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/AddEmbeddingsRequest.java]: good question:) changed

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java (hunk 30 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStore.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryResponse.java (hunk 35 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryResponse.java:8]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java (hunk 36 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/QueryRequest.java]: done

File: langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java (hunk 34 above)
Comment (kuraleta) [reply langchain4j/src/main/java/dev/langchain4j/store/embedding/chroma/CollectionCreationRequest.java]: ok, done

**Review by ghost — APPROVED** [review]
Good job!

**Pull request merged**